### PR TITLE
Add de.bforartists.Bforartists

### PR DESCRIPTION
**Pull request by StandingPadAnimations**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [ ] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
~~- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/Bforartists/Bforartists/issues/3614~~
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission

EDIT: added link regarding upstream discussion about a Bforartists flatpak

EDIT 2:
Regarding upstream: https://github.com/flathub/flathub/pull/4295#issuecomment-1617478647

Permissions have been changed for filesystem to now be only `home` and `/tmp`. `/tmp` is needed to prevent `/app` from being filled with gigabytes of autosaves, which has been found to prevent other Flatpaks from running.

EDIT 3: I have been banned from upstream, not sure what happens now

### CODE REVIEW

**Comment by Draise14** [comment]
<3

**Comment by hfiguiere** [comment]
there is already https://github.com/flathub/flathub/pull/3357

File: de.bforartists.Bforartists.json [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            "autodelete": true
+        },
+        "org.freedesktop.Platform.ffmpeg-full": {
+            "directory": "lib/ffmpeg",
+            "version": "22.08",
+            "add-ld-path": "."
+        }
+    },
+    "cleanup-commands": [
+        "mkdir -p /app/lib/ffmpeg"
```
Comment (hfiguiere) [inline de.bforartists.Bforartists.json:34]: cleanup is to cleanup, not to create.

Please add this to the post-install of the main module.

**Comment by StandingPadAnimations** [comment]
> there is already #3357

#3357 doesn't involve upstream though. This version was created with permission from upstream

File: de.bforartists.Bforartists.json [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                {
+                    "type": "archive",
+                    "url": "https://github.com/Bforartists/Bforartists/releases/download/v3.5.1/Bforartists-3-5-1-Linux.tar.xz",
+                    "sha256": "a179a87cd295c7d1df4aad0aab4c4a2b14665aa21fd2e822b3583bfba6ed5b6a",
+                    "strip-components": 0 
+                },
+                {
+                    "type": "shell",
+                    "commands": [
+                        "mv Bforartists*Linux* bforartists"
```
Comment (hfiguiere) [inline de.bforartists.Bforartists.json:156]: this should just be in the build-commands.

**Comment by StandingPadAnimations** [comment]
Alright, I've made the suggested changes

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

**Comment by hfiguiere** [comment]
the appstream file is missing. either not installed, in the wrong location or with the wrong name.

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

File: de.bforartists.Bforartists.json [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            "build-commands": [
+                "mv Bforartists*Linux* bforartists",
+                "install -Dm755 bforartists.sh /app/bin/bforartists",
+                "mv bforartists /app/bforartists",
+                "install -Dm644 /app/bforartists/bforartists.desktop /app/share/applications/$FLATPAK_ID.desktop",
+                "desktop-file-edit --set-icon=${FLATPAK_ID} /app/share/applications/${FLATPAK_ID}.desktop",
+                "install -Dm644 /app/bforartists/bforartists.svg /app/share/icons/hicolor/scalable/apps/$FLATPAK_ID.svg",
+                "install -Dm644 /app/bforartists/bforartists-symbolic.svg /app/share/icons/hicolor/symbolic/apps/$FLATPAK_ID-symbolic.svg",
+                "cd /app/bforartists/*.*/datafiles && mv locale /app/share/locale && ln -sf /app/share/locale locale", 
+                "install -Dm644 $FLATPAK_ID.appdata.xml /app/share/metainfo/$FLATPAK_ID.metainfo.xml"
```
Comment (hfiguiere) [inline de.bforartists.Bforartists.json:143]: this file isn't listed in the manifest.

File: de.bforartists.Bforartists.json [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        "--env=TMP=/tmp"
+    ],
+    "add-extensions": {
+        "de.bforartists.bforartists.Codecs": {
+            "directory": "lib/codecs",
+            "add-ld-path": "lib",
+            "bundle": true,
+            "autodelete": true
+        },
+        "org.freedesktop.Platform.ffmpeg-full": {
```
Comment (hfiguiere) [inline de.bforartists.Bforartists.json:27]: not sure if you need this since you build ffmpeg.

File: de.bforartists.Bforartists.metainfo.xml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      Bforartists is a fork of the popular open source 3D software Blender. Developed by some 3D enthusiasts to make Blender&apos;s UI even better.
+    </p>
+  </description>
+  
+  <launchable type="desktop-id">bforartists.desktop</launchable>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://www.bforartists.de/wp-content/uploads/2020/04/modelingbfa2-1024x619.jpg</image>
+    </screenshot>
+  </screenshots>
```
Comment (hfiguiere) [inline de.bforartists.Bforartists.metainfo.xml:22]: missing mandatory `url`, `content_rating`, `releases`.

File: de.bforartists.Bforartists.metainfo.xml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  <metadata_license>CC-BY-4.0</metadata_license>
+  <project_license>GPL-2.0-or-later</project_license>
+  
+  <description>
+    <p>
+      Bforartists is a fork of the popular open source 3D software Blender. Developed by some 3D enthusiasts to make Blender&apos;s UI even better.
+    </p>
+  </description>
+  
+  <launchable type="desktop-id">bforartists.desktop</launchable>
```
Comment (hfiguiere) [inline de.bforartists.Bforartists.metainfo.xml:17]: this should match the installed desktop file name.

File: de.bforartists.Bforartists.metainfo.xml
Comment (hfiguiere) [inline de.bforartists.Bforartists.metainfo.xml:1]: ideally this file should be shipped with the release tarball in a future release.

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

File: de.bforartists.Bforartists.json [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                    "type": "script",
+                    "commands": [
+                        "cd /app/bforartists",
+                        "exec ./bforartists \"$@\""
+                    ],
+                    "dest-filename": "bforartists.sh"
+                }
+            ],
+            "post-install": [
+                "mkdir -p /app/lib/ffmpeg"
```
Comment (hfiguiere) [inline de.bforartists.Bforartists.json:161]: I don't think this is needed anymore since you removed the ffmpeg extension.

**Comment by StandingPadAnimations** [comment]
I've been doing some travel the past couple of hours so I haven't been able to build locally all day. I've now built locally and it works, so fingers crossed it works here

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

**Comment by StandingPadAnimations** [comment]
Just curious, are there any additional things for me to change?

**Comment by barthalion** [comment]
Reviews and merges happen on the best effort basis, otherwise I would be reading YAMLs and JSONs 24/7.

File: de.bforartists.Bforartists.json [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        "--env=TMP=/tmp"
+    ],
+    "add-extensions": {
+        "de.bforartists.bforartists.Codecs": {
+            "directory": "lib/codecs",
+            "add-ld-path": "lib",
+            "bundle": true,
+            "autodelete": true
+        }    
+    },
```
Comment (barthalion) [inline de.bforartists.Bforartists.json:27]: What's the point of that? There's no reason to split codecs to separate extension, unless explicitly requested by some US-based entity interested in shipping this app by default with their distro. Not a blocker, just curious why you did that.

File: de.bforartists.Bforartists.json [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                "desktop-file-edit --set-icon=${FLATPAK_ID} /app/share/applications/${FLATPAK_ID}.desktop",
+                "install -Dm644 /app/bforartists/bforartists.svg /app/share/icons/hicolor/scalable/apps/$FLATPAK_ID.svg",
+                "install -Dm644 /app/bforartists/bforartists-symbolic.svg /app/share/icons/hicolor/symbolic/apps/$FLATPAK_ID-symbolic.svg",
+                "cd /app/bforartists/*.*/datafiles && mv locale /app/share/locale && ln -sf /app/share/locale locale", 
+                "install -Dm644 $FLATPAK_ID.metainfo.xml /app/share/metainfo/$FLATPAK_ID.metainfo.xml"
+            ],
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "https://github.com/Bforartists/Bforartists/releases/download/v3.5.1/Bforartists-3-5-1-Linux.tar.xz",
```
Comment (barthalion) [inline de.bforartists.Bforartists.json:143]: Could you configure https://github.com/flathub/flatpak-external-data-checker/ for that?

File: de.bforartists.Bforartists.json (hunk 8 above)
Comment (StandingPadAnimations) [reply de.bforartists.Bforartists.json:27]: It was found for the Blender flatpak that Freedesktop's FFMPEG doesn't play well with Blender (namely OpenH264). Bforartists is a Blender fork with UI changes, so what affects Blender also affects Bforartists

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

**Comment by StandingPadAnimations** [comment]
I've restricted the filesystem permissions to `home` and `/tmp`. `/tmp` as it was found in the Blender Flatpak that autosaves take up most of the space in `/run`, preventing other Flatpaks from running (https://github.com/flathub/org.blender.Blender/issues/122)

**Comment by StandingPadAnimations** [comment]
For readability reasons, I've converted the manifest from JSON to YAML (using yq). Currently testing to see if the Flatpak still builds locally

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

**Comment by StandingPadAnimations** [comment]
Update regarding upstream: it seems that they don't want to cooperate with discussing about Flatpak, calling it "propoganda". All I have done is give the reasons why I think a Flatpak would be beneficial, and give a set of comparisons regarding installation between Flatpak and Appimage. For that comparision in particular, I made it clear I was not continuing the Flatpak vs Appimage debate and was merely showing a comparison (I also made it clear that Appimage only needed a manager for desktop integration, not Appimage as a whole). The comparison was deleted by upstream on the basis of being propoganda. As such, here's a screenshot of my current comment just in case upstream decides to delete it:

Screenshot_20230703_014430_Mull

I'm not sure how this affects this submission, but I do know Bforartists is licensed under GPL (as it's a Blender fork, and Blender is licensed under GPL), so I'll continue exercising my right under GPL. In addition, I will continue to comply with reasonable requests from upstream. 

In the end, I believe I have been as civil as possible in the discussion with upstream (and frankly will continue to be civil with them), but like I said, I don't know how much that'll have an impact here.

**Comment by StandingPadAnimations** [comment]
If y'all want, I can close this submission since efforts to work with upstream haven't gone well

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

**Comment by StandingPadAnimations** [comment]
And here goes upstream deleting all of the comments on the tracker upstream https://github.com/Bforartists/Bforartists/issues/3614#issuecomment-1619541489

**Comment by StandingPadAnimations** [comment]
I'm banned upstream as well... not surprised. 

@hfiguiere @barthalion I'm honestly not sure what to do in light of what has happened upstream, do I continue on the Flatpak or would it be best if I close this PR?

File: de.bforartists.Bforartists.yaml [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+      - install -Dm644 $FLATPAK_ID.metainfo.xml /app/share/metainfo/$FLATPAK_ID.metainfo.xml
+
+    # Bforartists tar archive, retrived from
+    # GitHub releases and verified with a sha256 
+    # checksum. For every update, the URL and checksum 
+    # need to be updated. For the checksum, it is as simple
+    # as downloading the tar archive and running
+    # sha256sum <bforartists tar archive>
+    #
+    # In addition, the version for x-checker-data needs be updated
```
Comment (barthalion) [inline de.bforartists.Bforartists.yaml:150]: The point of x-checker-data is to save you the trouble and have something else detect the new version.

Comment (StandingPadAnimations) [reply de.bforartists.Bforartists.yaml:150]: So something like this?
```yaml
x-checker-data:
  type: anitya
  project-id: 359756
  stable-only: true
  versions:
    '>=': '3.5.1'
  url-template: https://github.com/Bforartists/Bforartists/releases/download/v$version/Bforartists-$version0-$version1-$version2-Linux.tar.xz
```

**Comment by hfiguiere** [comment]
bot, build de.bforartists.Bforartists

File: de.bforartists.Bforartists.yaml [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        # In addition, this also checks for new updates 
+        # to the tar archive automatically
+        x-checker-data:
+          type: anitya
+          project-id: 359756
+          stable-only: true
+          versions:
+            '>=': '3.5.1'
+          url-template: https://github.com/Bforartists/Bforartists/releases/download/v$version/Bforartists-$version0-$version1-$version2-Linux.tar.xz
+
```
Comment (barthalion) [inline de.bforartists.Bforartists.yaml:169]: Not sure where these version0, 1, 2 would come from? Docs mention `$version` only: https://github.com/flathub/flatpak-external-data-checker/#anitya-release-monitoring-checker

JSON checker offers more flexibility as it runs jq: https://github.com/flathub/flatpak-external-data-checker/#json-checker

Comment (StandingPadAnimations) [reply de.bforartists.Bforartists.yaml:169]: `version0`, `1`, and  `2` come from the URL checker, which anitya allows: https://github.com/flathub/flatpak-external-data-checker/#url-templates

However, JSON does look like less of a headache to do

**Comment by StandingPadAnimations** [comment]
bot, build de.bforartists.Bforartists

**Pull request closed without merging**